### PR TITLE
Support for installing add-ons from password protected registries

### PR DESCRIPTION
**Pull request by skateman**
I started playing around with private docker registries and realized that the supervisor can't deal with addons living in them. In my specific case the problem is that the GitHub Packages Docker registry cannot be accessed anonymously.

The idea here is to have a docker client's configfile under `HASSIO_DATA/docker.json`. The registries specified in this file are iterated through and the docker client authenticates against each of these. This allows us to pull addons from password-protected registries.

https://community.home-assistant.io/t/add-ons-from-password-protected-docker-registries/225746

closes https://github.com/home-assistant/supervisor/issues/2039

### CODE REVIEW

File: tests/docker/test_login.py
Comment (skateman) [inline tests/docker/test_login.py:30]: The reason behind this test is that if the mounted file doesn't exist on the host, docker creates a folder in the target container with that name.

**Comment by pvizeli** [comment]
We have here 2 layers:
- OS need store the credential and manage this
- Supervisor can manage the logins

We can provide an API endpoint which allow user to login, but we don't need store this data. On OS we need make sure that docker can write this credential persists and they get snapshot with RFC https://github.com/home-assistant/supervisor/issues/1866

**Comment by pvizeli** [comment]
Otherwise it's difficult to restore addon if we not directly know if we can pull the image. If we want handling this on Supervisor:

- create docker/login.py with a login manager and use our json logic and attach it to DockerAPI
- create an API endpoints to manage that secrets
- on every add-on install, let's give this manager the registry url and he make sure they is logged in

We don't map files into container, all need following the same logic and exposed to API

**Comment by skateman** [comment]
@pvizeli I tried to implement the stuff based on your recommendations, but some stuff weren't clear and I was improvising a lot by looking at existing pieces of code. So I guess there will be some issues and the tests are missing for now, but it does what you described and it's ready for a round of review.

File: supervisor/docker/validate.py
Comment (pvizeli) [inline supervisor/docker/validate.py:16]: Make just one schema

File: supervisor/docker/interface.py
Comment (pvizeli) [inline supervisor/docker/interface.py:99]: Every image has an `/` in, but you can use the regex from validation now. Please offload the part above into an internal function `_docker_login` on the interface.

File: supervisor/api/docker.py
Comment (pvizeli) [inline supervisor/api/docker.py:32]: Define the schema in that file, because the file or API can change over time and go forget where it was used. You can look to other API parts how they do it.

**Review by pvizeli — CHANGES_REQUESTED** [review]
Looks fine. Some small adjustments, after that we can merge it 👍

**Comment by pvizeli** [comment]
We need to get this data also into a snapshot and restore it

File: supervisor/docker/validate.py
Comment (skateman) [reply supervisor/docker/validate.py:8]: I'm not entirely sure what does this do :confused:

File: supervisor/docker/interface.py
Comment (skateman) [reply supervisor/docker/interface.py:99]: So I guess it should be a regex that matches at least two `/` in a string, right? Should the condition also go to the `_docker_login` or just anything below it?

**Comment by skateman** [comment]
@pvizeli I had some questions above, can you please answer them? Thanks!

File: supervisor/docker/interface.py
Comment (pvizeli) [reply supervisor/docker/interface.py:99]: The regex they we use for config validation can be use here to detect an host inside image url

File: supervisor/docker/validate.py
Comment (pvizeli) [reply supervisor/docker/validate.py:8]: https://github.com/home-assistant/supervisor/blob/dev/supervisor/validate.py#L147-L161

File: supervisor/api/docker.py
Comment (pvizeli) [reply supervisor/api/docker.py:32]: https://github.com/home-assistant/supervisor/blob/dev/supervisor/api/supervisor.py#L55-L70

File: supervisor/docker/validate.py
Comment (skateman) [reply supervisor/docker/validate.py:8]: I mean what does the empty regex should do here...

**Comment by skateman** [comment]
> We need to get this data also into a snapshot and restore it

Gonna figure out how to do this, the rest should be done, except I still don't understand [this](https://github.com/home-assistant/supervisor/pull/2038#discussion_r488608761) :confused:

File: supervisor/api/docker.py
Comment (pvizeli) [inline supervisor/api/docker.py:18]:
```suggestion
            vol.Required(ATTR_USERNAME): str,
            vol.Required(ATTR_PASSWORD): str,
```

File: supervisor/docker/validate.py
Comment (pvizeli) [inline supervisor/docker/validate.py:8]:
```suggestion
RE_REGISTRY = re.compile(r"^([a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,}$")
```

File: supervisor/docker/interface.py
Comment (pvizeli) [inline supervisor/docker/interface.py:21]:
```suggestion
IMAGE_WITH_HOST = re.compile(r"^((?:[a-z0-9]+(?:-[a-z0-9]+)*\.)+[a-z]{2,})/.+")
```

**Review by pvizeli — CHANGES_REQUESTED** [review]
Looks good, just fix the snapshot part and it's ready to merge.

1) Update snapshot schema.
2) Decrypt/Encrypt passwort like HA token

File: supervisor/snapshots/__init__.py
Comment (ludeeus) [inline supervisor/snapshots/__init__.py:305]: Why only restore this `if FOLDER_HOMEASSISTANT in folders` ?

File: supervisor/snapshots/__init__.py
Comment (ludeeus) [inline supervisor/snapshots/__init__.py:244]: This should be done before the restore of repositories

File: supervisor/snapshots/snapshot.py
Comment (ludeeus) [inline supervisor/snapshots/snapshot.py:141]: This will throw a `KeyError` if  `ATTR_DOCKER` is not in the snapshot?

File: supervisor/snapshots/snapshot.py
Comment (ludeeus) [inline supervisor/snapshots/snapshot.py:519]: This will throw a `TypeError` if  `self.docker` is not a dict (`None` if not set)?

**Review by ludeeus — COMMENTED** [review]
Added a few comments, please don't force push after review has started, it makes it hard to follow the process :+1:

File: supervisor/snapshots/snapshot.py
Comment (pvizeli) [inline supervisor/snapshots/snapshot.py:504]: Should we also encrypt the username?

**Review by pvizeli — APPROVED** [review]
Ready to merge after @ludeeus commits are addressed

File: supervisor/snapshots/snapshot.py
Comment (ludeeus) [reply supervisor/snapshots/snapshot.py:504]: In what scenario would that be needed?

Comment (pvizeli) [reply supervisor/snapshots/snapshot.py:504]: just a Suggestion

File: supervisor/snapshots/__init__.py
Comment (skateman) [reply supervisor/snapshots/__init__.py:305]: Should this be restored unconditionally? This is not a folder, but a file, so I can't just pass it to `folders` I guess. I could add a new attribute that would list all the configfiles we want to restore (and back up), but I am afraid that would be a more invasive change than this whole PR itself.

File: supervisor/snapshots/snapshot.py
Comment (skateman) [reply supervisor/snapshots/snapshot.py:504]: So should this be encrypted as well? TBH I don't see why, but your call...

Comment (ludeeus) [reply supervisor/snapshots/snapshot.py:504]: I think it's fine to skip it for now, we can always add that later if needed

File: supervisor/snapshots/__init__.py
Comment (ludeeus) [reply supervisor/snapshots/__init__.py:305]: Yeah, we should always store and restore this configuration, and the restoration of it needs to happen before the restoration of anything else container-related.

File: supervisor/snapshots/snapshot.py
Comment (ludeeus) [inline supervisor/snapshots/snapshot.py:141]: This returns a bool

```suggestion
        return self._data.get(ATTR_DOCKER, {})
```

File: supervisor/snapshots/__init__.py
Comment (ludeeus) [inline supervisor/snapshots/__init__.py:240]: Since we support custom image for homeassistant, move this to before the homeassistant restore

File: supervisor/snapshots/__init__.py
Comment (ludeeus) [inline supervisor/snapshots/__init__.py:323]: Move this so it's the first action under `async with snapshot:`

File: supervisor/snapshots/__init__.py
Comment (ludeeus) [reply supervisor/snapshots/__init__.py:240]: restore_dockerconfig is not awaitable

**Review by ludeeus — CHANGES_REQUESTED** [review]

**Comment by skateman** [comment]
@ludeeus thanks for your help, I fixed the issues. Should I squash my commits before merging, or is it okay as it is?

**Comment by ludeeus** [comment]
We squash all PRs on merge, so no need for you to do that in your branch 👍

File: supervisor/docker/__init__.py
Comment (ludeeus) [inline supervisor/docker/__init__.py:85]:
```suggestion
        return self._data.get(ATTR_REGISTRIES, {})
```

File: API.md
Comment (ludeeus) [inline API.md:1240]:
```suggestion
source when pulling docker images.
```

**Review by ludeeus — CHANGES_REQUESTED** [review]

**Review by ludeeus — APPROVED** [review]
Looks good :+1: 
Since there are no tests here, have you tested the current code?

**Comment by skateman** [comment]
@ludeeus I did, but I will run one more round with the edge cases you discovered, e.g. old snapshot without a docker config. Give me like an hour...

**Comment by skateman** [comment]
I downloaded the HassOS VM and went with `docker exec` into the supervisor, installed the `patch` command, applied the changes from this PR and ran the `compileall` command I found in the Dockerfile. Then outside the container I ran `docker commit` and `docker tag` on the supervisor image so both the `latest` and `245` had my changes. I immediately created a snapshot after the restart and also between the following steps. I added a [new addon repository](https://github.com/skateman/hh-hassio-repo/) and set its credentials using the API. I was able to install both packages from the repo and I was able to restore back and forth any snapshot and the changes were always there without any error.

**Comment by ludeeus** [comment]
Nice work @skateman! :medal_sports:

**Pull request merged**
